joern-cli/frontends/c2cpg/src/main/scala/io/joern/c2cpg/astcreation/FullNameProvider.scala

```diff
@@ -0,0 +1,337 @@
+      case f: ICPPASTFunctionDefinition   => shortNameForICPPASTFunctionDefinition(f)
+      case f: IASTFunctionDefinition      => shortNameForIASTFunctionDefinition(f)
+      case d: CPPASTIdExpression          => shortNameForCPPASTIdExpression(d)
+      case d: IASTIdExpression            => lastNameOfQualifiedName(ASTStringUtil.getSimpleName(d.getName))
+      case u: IASTUnaryExpression         => shortName(u.getOperand)
+      case c: IASTFunctionCallExpression  => shortName(c.getFunctionNameExpression)
+      case s: IASTSimpleDeclSpecifier     => s.getRawSignature
+      case e: IASTEnumerationSpecifier    => ASTStringUtil.getSimpleName(e.getName)
+      case c: IASTCompositeTypeSpecifier  => ASTStringUtil.getSimpleName(c.getName)
+      case e: IASTElaboratedTypeSpecifier => ASTStringUtil.getSimpleName(e.getName)
+      case s: IASTNamedTypeSpecifier      => ASTStringUtil.getSimpleName(s.getName)
+      case l: ICPPASTLambdaExpression     => nextClosureName()
+      case other =>
+        notHandledYet(other)
+        nextClosureName()
+    }
+    StringUtils.normalizeSpace(name)
+  }
+
+  private def fullNameFromBinding(node: IASTNode): Option[String] = {
+    node match {
+      case declarator: CPPASTFunctionDeclarator =>
+        declarator.getName.resolveBinding() match {
+          case function: ICPPFunction =>
+            val fullNameNoSig = function.getQualifiedName.mkString(".")
+            val fn = if (function.isExternC) {
+              function.getName
+            } else {
+              s"$fullNameNoSig:${functionTypeToSignature(function.getType)}"
+            }
+            Option(fn)
+          case x @ (_: ICPPField | _: CPPVariable) =>
+            val fullNameNoSig = x.getQualifiedName.mkString(".")
+            val fn = if (x.isExternC) {
+              x.getName
+            } else {
+              s"$fullNameNoSig:${cleanType(safeGetType(x.getType))}"
+            }
+            Option(fn)
+          case _: IProblemBinding =>
+            val fullNameNoSig = ASTStringUtil.getQualifiedName(declarator.getName)
+            val fixedFullName = fixQualifiedName(fullNameNoSig).stripPrefix(".")
+            if (fixedFullName.isEmpty) {
+              Option(s"${X2CpgDefines.UnresolvedNamespace}:${X2CpgDefines.UnresolvedSignature}")
+            } else {
+              Option(s"$fixedFullName:${X2CpgDefines.UnresolvedSignature}")
+            }
+          case _ => None
+        }
+      case declarator: CASTFunctionDeclarator =>
+        declarator.getName.resolveBinding() match {
+          case cVariable: CVariable => Option(cVariable.getName)
+          case _                    => Option(declarator.getName.toString)
+        }
+      case definition: ICPPASTFunctionDefinition =>
+        Option(fullName(definition.getDeclarator))
+      case _ => None
+    }
+  }
+
+  private def fullNameForCPPASTIdExpression(d: CPPASTIdExpression): String = {
```

Comment:
This method is called as fallback if fullNameFromBinding returns None. But it itself tries to work on bindings again. Does not seem to make sense.